[SystemZ][z/OS] Implement #pragma export

perry-ca opened the pull request:
Implement the export pragma that is used in the z/OS XL C/C++ compiler to indicate that an external symbol is to be exported from the shared library.  The syntax for the pragma is:
```
'#pragma' 'export' '(' name ')'
```
For C++ if `name` is a function it needs to be declared `extern "C"`.

See the following for the XL documentation:
- https://www.ibm.com/docs/en/zos/3.1.0?topic=descriptions-pragma-export

This code was originally in PR https://github.com/llvm/llvm-project/pull/111035. I have split it out into separate PRs so the code for #pragma export is in one PR and the code for _Export keyword is in another. See that original PR for earlier comments.

perry-ca committed:
#pragma export support

perry-ca committed:
Merge remote-tracking branch 'origin/main' into perry/pragma-export-only

erichkeane commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -918,6 +918,11 @@ WebAssembly Support
+SystemZ Support
+^^^^^^^^^^^^^^^
+
+- Add support for `#pragma export` for z/OS
```
We definitely need a much better intro/documentation here in the release notes.

erichkeane commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -2278,6 +2278,39 @@ class Sema final : public SemaBase {
+  /// haven't seen the declaration for yet. The TypeList is the argument list
+  /// the function must match if HasTypeList is true.
+  struct SymbolLabel {
+    std::optional<SmallVector<QualType, 4>> TypeList;
```
```suggestion
    std::optional<SmallVector<QualType>> TypeList;
```

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -568,6 +574,11 @@ void Parser::initializePragmaHandlers() {
   MaxTokensTotalPragmaHandler = std::make_unique<PragmaMaxTokensTotalHandler>();
   PP.AddPragmaHandler("clang", MaxTokensTotalPragmaHandler.get());
 
+  if (getLangOpts().ZOSExt) {
```
This isn't particularly ZOS specific other than for compiler compat, right?  Is there a good reason we couldn't allow this syntax on other compilers?

erichkeane replied:
IDK why I didn't see the comment response here... I think doing these based on the target here is somewhat reasonable, it harms portability of this attribute, but in a way that I don't think is particularly challenging thanks to the ZOSExt flag being required.

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1395,6 +1411,171 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+                           const IdentifierInfo *IdentName) {
+  NestedNameSpecifier *NestedId = nullptr;
+  if (PP.getLangOpts().CPlusPlus) {
+    if (Tok.is(tok::coloncolon)) {
```
Presumably this should ensure we only look up in the global namespace, right?  VS local namespace?

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1395,6 +1411,171 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+            << PragmaName;
+        return nullptr;
+      }
+      IdentifierInfo *II = Tok.getIdentifierInfo();
```
Identifier isn't sufficient, what if this is naming a type inside a template?  Or a dependent type?

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1395,6 +1411,171 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+bool Parser::zOSParseParameterList(
+    StringRef PragmaName, std::optional<SmallVector<QualType, 4>> &TypeList,
+    Qualifiers &CVQual) {
+  if (Tok.is(tok::l_paren)) {
```
We probably want to use a BalancedDelimiterTracker/etc for the parens.

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1395,6 +1411,171 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+      if (!TResult.isInvalid()) {
+        QualType QT = TResult.get().get();
+        if (!QT.getTypePtr()->isVoidType()) {
+          TypeList->push_back(QT);
```
This feels lossy... Also, why are we allowing 'void' in this list at all?

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1395,6 +1411,171 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+    }
+
+    if (TypeList.has_value())
+      while (Tok.is(tok::kw_const) || Tok.is(tok::kw_volatile)) {
```
This part seems wrong too, what about const before?

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1395,6 +1411,171 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+      return false;
+    }
+  } while (Tok.isNot(tok::eof));
+  PP.Lex(Tok);
```
There is quite a bit on how htis parsing is working that doesn't seem right to me, but the parser code owner shoudl take a look.

perry-ca replied:
Can you take a look at the parsing again.  I have removed the part that parsed the parameter list as part of removing that form of the pragma syntax.

erichkeane reviewed (commented):
I only made it through the first bit of hte implementation, but parsing doesn't seem right to me, it doesn't recovery nicely as far as I can tell.

AS far as the design here, it isn't clear to me how this manages all of the C++ lookup/scoping/etc, and I don't see how the data structure provided would fix that.

Also, you seem to be doing some sort of parse during statement parsing, but no tests that show what this does/how this works when the scope isn't the local one?  Or when this is in a function definition/etc.

ALSO: I don't see any template instantiation code, which this would need for statement/decl level declaration.

perry-ca committed:
only accept export(id)

perry-ca committed:
Merge remote-tracking branch 'origin/main' into perry/pragma-export-only

perry-ca committed:
update release notes

perry-ca committed:
re-org the checks to simplify the logic

perry-ca committed:
more simplification

perry-ca committed:
remove extra field

perry-ca committed:
Merge branch 'llvm:main' into perry/pragma-export-only

perry-ca committed:
Merge remote-tracking branch 'origin/main' into perry/pragma-export-only

perry-ca committed:
staging

perry-ca committed:
staging

perry-ca committed:
staging

perry-ca committed:
Merge branch 'main' into perry/pragma-export-only

perry-ca committed:
NFC: formatting

perry-ca committed:
Merge remote-tracking branch 'origin/main' into perry/pragma-export-only

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1441,7 +1440,7 @@ bool Parser::zOSHandlePragmaHelper(tok::TokenKind PragmaKind) {
 
     PP.Lex(Tok);
-    Actions.ActOnPragmaExport(IdentName, IdentNameLoc);
+    Actions.ActOnPragmaExport(IdentName, IdentNameLoc, getCurScope());
```
Can `ActOnPragmaExport` not check the current scope?

perry-ca replied:
The current scope is available inside of the function.  However, the current scope inside of the function may not match the scope where the name reference occurred.  In this case the two are the same.  I thought about removing the argument but I noticed that it is common to pass the the current scope for these `ActOnPragma*` functions when the current scope is needed.  I'm leaning towards keeping the argument to be consistent.

erichkeane commented on clang/lib/Sema/SemaAttr.cpp:
```diff
@@ -1327,23 +1327,35 @@ void Sema::AddImplicitMSFunctionNoBuiltinAttr(FunctionDecl *FD) {
+        return D;
+    if (isa<VarDecl>(D->getCanonicalDecl()))
+      return D;
+  }
```
I don't think this lookup is quite right.  We should probably be just doing normal name lookup rather than this.  Also, this doesn't support dependent names, which we definitely should be doing.

perry-ca replied:
Anything in particular that doesn't look right?  The use of lookupName is similar to other uses.  This does have the loop after that looks for the extern "C" declaration.  That is part of the functionality of the pragma.

Dependent names aren't supported.  This is pragma can only be used at file scope.  There aren't any dependent names there.

erichkeane replied:
Clarified ab it above, not sure exactly what I was th inking before, but looked at with fresh eyes.

erichkeane reviewed (commented):
Missing a bunch of tests as well.  Did those disappear?  We should also probably have a document somewhere that explains exactly what this extension is doing, so that we can better understand how to review it.

perry-ca commented:
> I mentioned above, its missing a massive amount of dependent cases/etc. Also, the not-allowing-FQNs seems unfortunate, though you're at least testing those as errors.
> 
The pragma only supports external function and variable names and doesn't support fully qualified names.  I don't think you can have a valid case where the name is a dependent name.  Do you have an example of what you are thinking of?

> As far as documentation, we have LanguageExtensions.rst that I think this fits into.
I've got that written and will add the commit shortly.

erichkeane commented:
> > I mentioned above, its missing a massive amount of dependent cases/etc. Also, the not-allowing-FQNs seems unfortunate, though you're at least testing those as errors.
> 
> The pragma only supports external function and variable names and doesn't support fully qualified names. I don't think you can have a valid case where the name is a dependent name. Do you have an example of what you are thinking of?
> 
> > As far as documentation, we have LanguageExtensions.rst that I think this fits into.
> > I've got that written and will add the commit shortly.

Right, not supporting fully qualified names (for something like: `SomeNS::Foo::Bar`) is unfortunate.  

It looks like you're NOT supporting external functions, every example I see is a locally defined function?  At least in the tests?  They ALL do use C-linkage, but that is something entirely different.

Re dependent names... 

```
template<auto func> 
struct S {

#pragma export(func)
};

extern "C" void funcToExport();

S<funcToExport> s{};
```

Seems reasonable to me? Alternatively, cases where it would be referenced by the template argument.

perry-ca committed:
move check out of loop

perry-ca committed:
Merge remote-tracking branch 'origin/main' into perry/pragma-export-only

perry-ca committed:
Add documentation & enforce file scope usage

perry-ca committed:
Merge branch 'main' into perry/pragma-export-only

perry-ca committed:
Merge branch 'main' into perry/pragma-export-only

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1386,6 +1402,74 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+    if (Tok.isNot(tok::l_paren)) {
+      PP.Diag(Tok.getLocation(), diag::warn_pragma_expected_lparen)
+          << PragmaName;
+      return false;
```
Do you have any tests for recovering from parsing these 'nots'?  Typically you'd want to do a 'skip token's, else the parser is in a weird place.  Probably every where here this gives up.

erichkeane commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -1187,6 +1187,50 @@ specifier for ``_Float16``, and (unlike ``float``) it will not be implicitly pro
+  #pragma export (name)
+
+where ``name`` is the name of the external function or variable to be
+exported.  The symbol needs to have external linkage.  The pragma may appear
```
Ah, see it now, disregard questions on Function vs var.  The name of the `LookupExternC` should probably be renamed as it implies function, but you really mean 'or global variable'.

perry-ca commented on clang/lib/Sema/Sema.cpp:
```diff
@@ -1495,6 +1495,13 @@ void Sema::ActOnEndOfTranslationUnit() {
   }
 
+  // Visit all pending #pragma export.
+  for (auto &Iter : PendingExportedNames) {
```
no prob.  I would have thought it would have been considered obvious.  I see other examples of auto being used with DenseMap (eg. ExplicitlyMapped in clang/lib/Sema/SemaOpenMP.cpp).

perry-ca commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -7637,6 +7637,29 @@ static void emitReadOnlyPlacementAttrWarning(Sema &S, const VarDecl *VD) {
+    if (getLangOpts().CPlusPlus && !FD->isExternC())
+      return;
+  }
+  IdentifierInfo *IdentName = NewD->getIdentifier();
```
This is only ever called for functions and variables.  I'll add an assert to emphasis that.

perry-ca committed:
address review feedback

perry-ca commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1386,6 +1402,74 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+  ConsumeAnnotationToken();
+
+  do {
+    PP.Lex(Tok);
```
I took a look at this.  My conclusion is we should be calling PP.Lex().  The reasons are:
- ConsumeToken()/ConsumeAnyToken() track state information which applies to the user source code, not the pragmas.  An unbalanced '(' would affect code outside of the pragma
- All the other pragma handling functions call PP.Lex() too

perry-ca commented on clang/lib/Sema/SemaAttr.cpp:
```diff
@@ -1327,6 +1327,60 @@ void Sema::AddImplicitMSFunctionNoBuiltinAttr(FunctionDecl *FD) {
+    if (auto *FD = dyn_cast<FunctionDecl>(D->getCanonicalDecl()))
+      if (FD->isExternC())
+        return D;
+    if (isa<VarDecl>(D->getCanonicalDecl()))
```
The look up could match other declarations like typedefs.  This function shouldn't return those declarations, just functions and variables.  The name change helps clarify this.

perry-ca committed:
assert so user knows what is valid

perry-ca committed:
Merge branch 'main' into perry/pragma-export-only

erichkeane reviewed (commented):
Sorry for the delay!  I was out since the beginning of December and am finally back at my desk :) 

Down to just quick items I believe.

erichkeane commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -2835,6 +2835,15 @@ static void handleExternalSourceSymbolAttr(Sema &S, Decl *D,
       S.Context, AL, Language, DefinedIn, IsGeneratedDeclaration, USR));
 }
 
+void Sema::mergeVisibilityType(Decl *D, SourceLocation Loc,
```
Ah, hmm...  ok, i think we're ok with the visibility stuff staying around in this patch, it wasn't clear when I was reviewing it last week apparently.

perry-ca committed:
improve code quality

perry-ca committed:
rename SymbolLabel

perry-ca commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -2314,6 +2314,23 @@ class Sema final : public SemaBase {
+
+  /// A label from a C++ #pragma export, for a symbol that we
+  /// haven't seen the declaration for yet.
+  struct SymbolLabel {
```
I renamed the class `PendingPragmaInfo`.  I left "export" out of the name as I have another pragma (called map) that I'm going to upstream after this one.  Let me know if you're ok with that.

perry-ca commented:
@erichkeane I've addressed your comments.  I hope everything is good now.  I also changed `zOSPragmaHandlerHelper` so it skips to the end of the line instead of returning false and expecting the caller to.

perry-ca committed:
Merge branch 'main' into perry/pragma-export-only

erichkeane commented on clang/lib/Parse/ParsePragma.cpp:
```diff
@@ -1386,6 +1402,77 @@ bool Parser::HandlePragmaMSAllocText(StringRef PragmaName,
+                      false);
+  ConsumeAnnotationToken();
+
+  auto SkipToEnd = [this]() {
```
Instead of calling this everywhere, we can just do this as a :

`auto exit = llvm::make_scope_exit(<lambda>);`.  


That said, I THINK the `StopBeforeMatch` is wrong here.  In the `correct` case you consume the EOF, in the `error` case you're not?  Perhaps this should consume it too?

erichkeane reviewed (approved):
1 concern on how we are handling error recovery during parsing, else LGTM.

perry-ca committed:
add another error recovery test

perry-ca committed:
use scope_exit

perry-ca committed:
scope_exit will consume the EOF

perry-ca committed:
formating

perry-ca commented:
Is everything good now @erichkeane ?  Thanks for reviewing.

perry-ca committed:
switch back to PP.Lex()

perry-ca committed:
Merge branch 'main' into perry/pragma-export-only

perry-ca committed:
Merge branch 'main' into perry/pragma-export-only

perry-ca committed:
Merge branch 'main' into perry/pragma-export-only

perry-ca committed:
Merge remote-tracking branch 'origin/main' into perry/pragma-export-only
